rust/connlib/tunnel/src/client.rs

```diff
@@ -281,6 +291,22 @@ impl ClientState {
                 channel: p.channel.clone(),
             })
         }) else {
+            match self
+                .gateway_awaiting_connection_timers
+                // Note: we don't need to set a timer here because
+                // the FutureMap already expires things, it seems redundant
+                // to also have timer that expires.
+                .try_push(gateway, std::future::pending())
```

Comment:
@thomaseizinger this feels like I'm abusing the `FutureMap` interface, I'm open to other suggestions if this is not ideal

Reply:
Ha, neat workaround :D

Maybe a basic `SelectAll` with a `BoxFuture` that first sleeps and then emits the ID is a better design here?

Reply:
`JoinSet` seems to have done the trick :) though we lose the bounded part which was nice but it's a good tradeoff

Reply:
Note that the semantics are slightly different. A `JoinSet` seems to run things in the background by spawning them onto the runtime.

Reply:
ah, yeah, that makes sense, I think it works fine for this case and the types are a bit simpler to write

Reply:
I thought a bit more about this and I think I will go with the `FutureMap`, there's a *very* unlikely edge case that this prevents:
* This timer could have been initiated and this would delete the `gateway_id` from `gateway_awaiting_connection`
* We can lose connection and restart a connection intent to the same gateway
* The timer triggers while we are establishing a connection

with `FutureMap` we can remove the timer before it ever goes off if a connection is established, without an additional reference to the futures.